frame/beefy/src/lib.rs

```diff
@@ -105,20 +105,20 @@ impl<T: Config> Pallet<T> {
 	}
 
 	fn change_authorities(new: Vec<T::BeefyId>, queued: Vec<T::BeefyId>) {
-		// As in GRANDPA, we trigger a validator set change only if the the validator
-		// set has actually changed.
-		if new != Self::authorities() {
-			<Authorities<T>>::put(&new);
-
-			let next_id = Self::validator_set_id() + 1u64;
-			<ValidatorSetId<T>>::put(next_id);
-			if let Some(validator_set) = ValidatorSet::<T::BeefyId>::new(new, next_id) {
-				let log = DigestItem::Consensus(
-					BEEFY_ENGINE_ID,
-					ConsensusLog::AuthoritiesChange(validator_set).encode(),
-				);
-				<frame_system::Pallet<T>>::deposit_log(log);
-			}
+		// Always issue a change if `session` says that the validators have changed.
+		// Even if their session keys are the same as before, the underlying economic
+		// identities have changed. Furthermore, the digest below is used to signal
+		// BEEFY mandatory blocks.
+		<Authorities<T>>::put(&new);
+
+		let next_id = Self::validator_set_id() + 1u64;
+		<ValidatorSetId<T>>::put(next_id);
+		if let Some(validator_set) = ValidatorSet::<T::BeefyId>::new(new, next_id) {
+			let log = DigestItem::Consensus(
+				BEEFY_ENGINE_ID,
+				ConsensusLog::AuthoritiesChange(validator_set).encode(),
+			);
+			<frame_system::Pallet<T>>::deposit_log(log);
 		}
 
 		<NextAuthorities<T>>::put(&queued);
```

Comment:
Can we remove `<NextAuthorities<T>>` here and just track `<Authorities<T>>`?

I guess we need it because we are using new set for finalizing mandatory block, and need previous mandatory block's next_authority to validate the signers of commits in justification?
https://github.com/paritytech/substrate/pull/10882#pullrequestreview-894738340

If we can remove the `<NextAuthorities<T>>` concepts here, then some existing optimization like  `if new != Self::authorities()` could be possible again? 
(Currently, existing way is a bad optimization because it made `<NextAuthorities<T>>` mutable for specific beefy set id.)

Reply:
After [1085195](https://github.com/paritytech/substrate/pull/10882/commits/108519544b6e5767574b8c84bf5a34a5b27a8f8f) I think we can drop `<NextAuthorities<T>>` from the runtime api, yes. Will do in a later PR if nothing changes down the line.

> If we can remove the <NextAuthorities<T>> concepts here, then some existing optimization like if new != Self::authorities() could be possible again?

I don't understand this though, can you please elaborate?

Reply:
After [1085195](https://github.com/paritytech/substrate/pull/10882/commits/108519544b6e5767574b8c84bf5a34a5b27a8f8f), we will use `BeefyCurrentAuthoritySet` instead of `BeefyNextAuthoritySet`, and then we can remove `<NextAuthorities<T>>` storage from beefy pallet as you said, we now only track the current authorities.

One gas optimization we want to do in target chains (e.g. Ethereum) is, if we can do as little as possible mandatory header relay(transaction costs to Ethereum beefy finality contracts) if there no demand during this period and sessions, then we can save gas cost as a result.

Reduce the amount of unnecessary beefy authorities set changes by design is one way, and I guess this was why there was attempt do have `if new != Self::authorities()` in deprecated codes.  But when the set was trying to define underlying economic identities set with two parts (current_authorities, next_authorites), the optimization can not be applied by only checking `if new != Self::authorities()`, because `next_authorites` could also change even if `current_authorities` is not changed, enable this checking to set change will make the economic identities definition confuse and ambiguous.

If we removed the `next_authorites` from set definition, then I  think it make the gas optimization idea possible since from bridge point of view, we can have different beefy sets transitions rules independent of grandpa sets.

Reply:
> But ultimately, even when keys are reused, from the substrate-based-chain's point of view the session clearly changes, and BEEFY should also reflect that. I believe there is advantage in having predictable session lengths and that these match all other session-aware components in substrate.

Make sense.